File: solutions/2020-May-30-day-leetcoding-challenge/7.Cousins-in-Binary-Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
var isCousins = function (root, x, y) {
  const { val, left, right } = root;
  if (val == x || val == y) {
    return false;
  }

  const xDepth = findVal(root, left, x) || findVal(root, right, x);
  const yDepth = findVal(root, left, y) || findVal(root, right, y);

  return xDepth.parent != yDepth.parent && xDepth.depth == yDepth.depth;
};

function findVal(parent, child, target, depth = 1) {
  if (child == null) return null;

  const { val, left, right } = child;
  if (val == target) return { parent, depth };

  return (
    findVal(child, left, target, depth + 1) ||
    findVal(child, right, target, depth + 1)
  );
}

File: java/src/30/days/trail/2020/may/7.Cousins-in-Binary-Tree.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        var target1 = getDepth(root, root.left, x, 1);
        if (target1 == null) target1 = getDepth(root, root.right, x, 1);

        var target2 = getDepth(root, root.left, y, 1);
        if (target2 == null) target2 = getDepth(root, root.right, y, 1);

        if(target1 == null || target2 ==null) return false;

        return target1.left != target2.left && target1.val == target2.val;
    }

    TreeNode getDepth(TreeNode parent, TreeNode child, int target, int depth) {
        if (child == null) return null;

        if (child.val == target) {
            return new TreeNode(depth, parent, parent);
        }

        var result = getDepth(child, child.left, target, depth + 1);
        if (result != null) return result;
        return getDepth(child, child.right, target, depth + 1);
    }
}
